Add configuration point to SharedInformer to transform objects before storing

alexzielenski opened the pull request:
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:

We have teams who use client-go's `SharedIndexInformer` through controller-runtime extensively. Some applications create caches for object metadata. These caches can have tens of thousands of objects. We have found that some our controllers are being OOM-killed more frequently recently since the addition of `managedFields`.

This PR adds a configuration point to SharedInformer to arbitrarily transform the object before it is stored in the cache. We intend to use it to null out any large unused fields to save ram cost; since our controller cares only about a subset of the object's fields. It may also have other applications, as well.

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Adds SetTransform to SharedInformer to allow users to transform objects before they are stored.
```

alexzielenski commented:
/hold

I still need to add tests. Posting this PR in the meantime for comments.

wojtek-t commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -180,8 +180,17 @@ type SharedInformer interface {
+	// This function is intended for you to take the opportunity to
+	// remove, transform, or normalize fields. One use case is to strip unused
+	// metadata fields out of objects to save on RAM cost.
+	SetPreprocessHandler(handler CachePreprocessHandler) error
```
Can you unify your change with the TransformingInformer we already have:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/controller.go#L395

It seems you're trying to achieve pretty much the same thing.

alexzielenski replied:
@wojtek-t thanks for looking. I've modified the PR to refactor SharedIndexInformer to make use of the `newInformer` constructor rather than constructing its `controller` itself. Please take another look!

alexzielenski reviewed (commented):
I've modified the PR to use refactor `SharedIndexInformer` to now share code with `newInformer`. `SharedIndexInformer` has a few quirks which required some minor changes but it did fit fairly nicely.

There is one hairy area around sending the notification on update events. The `ResourceEventHandler` interface does not have an argument for the event type, so the original logic which silenced Sync notifications for non-sync listeners may be different and the new implementation needs to be looked at.

fedebongio commented:
/triage accepted

wojtek-t commented on staging/src/k8s.io/client-go/tools/cache/controller.go:
```diff
@@ -437,13 +440,20 @@ func newInformer(
+		ShouldResync:      shouldResync,
 
 		Process: func(obj interface{}) error {
+			if mutex != nil {
```
Instead of this, maybe we should just create a private

`processObject` top level function that is doing what it is currently doing (without your change) and use it here


Once done, you don't need to change anything else in this file. The boilerplate of creating a config is fine to be copied in the shared_informer. So it remains as it is now, with the only change of HandleDeltas calling this new processObject function.

alexzielenski replied:
@wojtek-t could you explain the benefit of this alternative way? I have no problem changing the PR, I would just like to understand.

wojtek-t replied:
It's subjective, but the more vars/params/etc. I have, the harder it is to follow them and ensure they work fine.
So I personally prefer to share the code that is complex, but sharding boilerplate (e.g. setting fields in the config) I don't think it brings much value. And complicates the code that was simpler before because wasn't using all of that.

alexzielenski replied:
@wojtek-t The commit has been updated with your suggestion

wojtek-t commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -534,49 +551,43 @@ func (s *sharedIndexInformer) AddEventHandlerWithResyncPeriod(handler ResourceEv
+	// If is a Sync event, isSync should be true
+	// If is a Replaced event, isSync is true if resource version is unchanged.
+	// If RV is unchanged, this is a Sync/Replaced event?
+
```
> I think this is semantically equivalent, but need someone with more kube experience to confirm.

for `Sync`, the RV by-definition has to be equal (object doesn't change), so this is fine
for `Replace` your check is equivalent

So the only missing bit is that for `Create` and `Update` RV has to change (so won't be caught by your logic)

So indeed - it seems equivalent.

alexzielenski committed:
modify SharedIndexInformer to use newInformer constructor which supports transformers

avoids code duplication, allows transformer to be used with SharedIndexInformer

alexzielenski commented:
/hold cancel

test added

wojtek-t commented:
/ok-to-test

/lgtm
/approve

wojtek-t commented:
/lgtm cancel
/approve cancel

Test failure is relevant to this PR - please fix.

alexzielenski commented:
Test failure was a bug in the test. Since the transformer is run even for relists/updates it is important for it copy before mutating, if it is going to change the object. Otherwise, there is a potential data race. I've corrected the test and added a note to `SetTransformer` to watch out for this.

alexzielenski commented:
@wojtek-t Tests passing again. Please re-lgtm

wojtek-t commented:
/lgtm
/approve

alexzielenski commented:
currently blocked by #107708

howardjohn commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -180,6 +181,20 @@ type SharedInformer interface {
+	//
+	// Must be set before starting the informer.
+	//
+	// Note: Since the object given to the handler may be already shared with
```
How does this happen? Seems like a sharedIndexInformer can only run once, and ProcessDeltas is called in controller.processLoop which is not concurrent.

DeepCopy may be expensive so given one of the primary use cases for this is performance it seems fairly important to not do extra work when not needed?

alexzielenski replied:
When a relist off existing data occurs it is possible for handledeltas to be called multiple times for the same object. The first time the transformation is applied it will not be concurrent. Assuming you have a way of marking an already- transformed object in order to skip repeated transformation applications then a copy should not be necessary.

howardjohn commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -180,6 +181,20 @@ type SharedInformer interface {
+	//	other goroutines, it is advisable to copy the object being
+	//  transform before mutating it at all and returning the copy to prevent
+	//	data races.
+	SetTransform(handler TransformFunc) error
```
I am a bit concerned that if an error occurs there is, as far as I can tell, no logging at all. This bubbles up to `processLoop`, which discards the error

howardjohn commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -180,6 +181,20 @@ type SharedInformer interface {
+
+	// The TransformFunc is called for each object which is about to be stored.
+	//
+	// This function is intended for you to take the opportunity to
```
It would be helpful to list the limitations of this. Does the returned type have to be the same type as the input?

From what I can tell it does not but must implement `runtime.Object` and `metav1.ObjectMetaAccessor` or equivalent interfaces

d3c3mber commented:
Just curious, why we are using the transform function in  `sharedIndexInformer.HandleDeltas()`, can we move the transformation function up to the reflector? @alexzielenski
